### PR TITLE
Add more tests for reduce

### PR DESCRIPTION
**Pull request by data-man**

### CODE REVIEW

**Comment by data-man** [comment]
Oops, test-aarch64-linux-musl:
LLVM ERROR: Cannot select: 0xffffca975da8: v2i64 = AArch64ISD::SMINV 0xffffcab3ad60, vector.zig:495:21

283 | 0xffffcab3ad60: v2i64 = insert_vector_elt 0xffffca522988, 0xffffca975658, Constant:i64<1>, vector.zig:495:21
284 | 0xffffca522988: v2i64 = scalar_to_vector 0xffffca9756c0, vector.zig:495:21
285 | 0xffffca9756c0: i64 = AArch64ISD::CSEL 0xffffca7e2de0, 0xffffca5224a8, Constant:i32<11>, 0xffffca9d0d98:1, vector.zig:495:21
286 | 0xffffca7e2de0: i64 = extract_vector_elt 0xffffcacadca0, Constant:i64<0>, vector.zig:495:21
287 | 0xffffcacadca0: v2i64,ch = load<(load 16 from %ir.4, align 8)> 0xffffca522850, 0xffffca522b90, undef:i64, vector.zig:495:33
288 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
289 | 0xffffcab3a3a0: i64 = Register %0
290 | 0xffffca975b38: i64 = undef
291 | 0xffffca9753e8: i64 = Constant<0>
292 | 0xffffca5224a8: i64 = extract_vector_elt 0xffffcda97b68, Constant:i64<0>, vector.zig:495:21
293 | 0xffffcda97b68: v2i64,ch = load<(load 16 from %ir.4 + 16, align 8)> 0xffffca522850, 0xffffca9758c8, undef:i64, vector.zig:495:33
294 | 0xffffca9758c8: i64 = add nuw 0xffffca522b90, Constant:i64<16>, vector.zig:495:33
295 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
296 | 0xffffcab3a3a0: i64 = Register %0
297 | 0xffffca523140: i64 = Constant<16>
298 | 0xffffca975b38: i64 = undef
299 | 0xffffca9753e8: i64 = Constant<0>
300 | 0xffffca522578: i32 = Constant<11>
301 | 0xffffca9d0d98: i64,i32 = AArch64ISD::SUBS 0xffffca7e2de0, 0xffffca5224a8, vector.zig:495:21
302 | 0xffffca7e2de0: i64 = extract_vector_elt 0xffffcacadca0, Constant:i64<0>, vector.zig:495:21
303 | 0xffffcacadca0: v2i64,ch = load<(load 16 from %ir.4, align 8)> 0xffffca522850, 0xffffca522b90, undef:i64, vector.zig:495:33
304 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
305 | 0xffffcab3a3a0: i64 = Register %0
306 | 0xffffca975b38: i64 = undef
307 | 0xffffca9753e8: i64 = Constant<0>
308 | 0xffffca5224a8: i64 = extract_vector_elt 0xffffcda97b68, Constant:i64<0>, vector.zig:495:21
309 | 0xffffcda97b68: v2i64,ch = load<(load 16 from %ir.4 + 16, align 8)> 0xffffca522850, 0xffffca9758c8, undef:i64, vector.zig:495:33
310 | 0xffffca9758c8: i64 = add nuw 0xffffca522b90, Constant:i64<16>, vector.zig:495:33
311 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
312 | 0xffffcab3a3a0: i64 = Register %0
313 | 0xffffca523140: i64 = Constant<16>
314 | 0xffffca975b38: i64 = undef
315 | 0xffffca9753e8: i64 = Constant<0>
316 | 0xffffca975658: i64 = AArch64ISD::CSEL 0xffffca51a798, 0xffffcab3a268, Constant:i32<11>, 0xffffca9d0988:1, vector.zig:495:21
317 | 0xffffca51a798: i64 = extract_vector_elt 0xffffcacadca0, Constant:i64<1>, vector.zig:495:21
318 | 0xffffcacadca0: v2i64,ch = load<(load 16 from %ir.4, align 8)> 0xffffca522850, 0xffffca522b90, undef:i64, vector.zig:495:33
319 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
320 | 0xffffcab3a3a0: i64 = Register %0
321 | 0xffffca975b38: i64 = undef
322 | 0xffffca522ed0: i64 = Constant<1>
323 | 0xffffcab3a268: i64 = extract_vector_elt 0xffffcda97b68, Constant:i64<1>, vector.zig:495:21
324 | 0xffffcda97b68: v2i64,ch = load<(load 16 from %ir.4 + 16, align 8)> 0xffffca522850, 0xffffca9758c8, undef:i64, vector.zig:495:33
325 | 0xffffca9758c8: i64 = add nuw 0xffffca522b90, Constant:i64<16>, vector.zig:495:33
326 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
327 | 0xffffcab3a3a0: i64 = Register %0
328 | 0xffffca523140: i64 = Constant<16>
329 | 0xffffca975b38: i64 = undef
330 | 0xffffca522ed0: i64 = Constant<1>
331 | 0xffffca522578: i32 = Constant<11>
332 | 0xffffca9d0988: i64,i32 = AArch64ISD::SUBS 0xffffca51a798, 0xffffcab3a268, vector.zig:495:21
333 | 0xffffca51a798: i64 = extract_vector_elt 0xffffcacadca0, Constant:i64<1>, vector.zig:495:21
334 | 0xffffcacadca0: v2i64,ch = load<(load 16 from %ir.4, align 8)> 0xffffca522850, 0xffffca522b90, undef:i64, vector.zig:495:33
335 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
336 | 0xffffcab3a3a0: i64 = Register %0
337 | 0xffffca975b38: i64 = undef
338 | 0xffffca522ed0: i64 = Constant<1>
339 | 0xffffcab3a268: i64 = extract_vector_elt 0xffffcda97b68, Constant:i64<1>, vector.zig:495:21
340 | 0xffffcda97b68: v2i64,ch = load<(load 16 from %ir.4 + 16, align 8)> 0xffffca522850, 0xffffca9758c8, undef:i64, vector.zig:495:33
341 | 0xffffca9758c8: i64 = add nuw 0xffffca522b90, Constant:i64<16>, vector.zig:495:33
342 | 0xffffca522b90: i64,ch = CopyFromReg 0xffffca465f48, Register:i64 %0
343 | 0xffffcab3a3a0: i64 = Register %0
344 | 0xffffca523140: i64 = Constant<16>
345 | 0xffffca975b38: i64 = undef
346 | 0xffffca522ed0: i64 = Constant<1>
347 | 0xffffca522ed0: i64 = Constant<1>
348 | In function: behavior.vector.S.doTheTestReduce.656

File: test/stage1/behavior/vector.zig [hunk 1]
```diff
@@ -495,75 +495,109 @@ test "vector reduce operation" {
             var r = @reduce(op, @as(Vector(N, TX), x));
             switch (@typeInfo(TX)) {
                 .Int, .Bool => expectEqual(expected, r),
-                .Float => {
```
Comment (LemonBoy) [inline test/stage1/behavior/vector.zig:498]: `expectWithinEpsilon` is dumb as a brick, that's why you need this to deal with NaNs.
If you don't believe me try changing the expected value of some nan-returning reduction, you'll see that the test will pass with no problem at all.

**Comment by LemonBoy** [comment]
> Oops, test-aarch64-linux-musl:

Already fixed in LLVM master, no idea what commit fixed it though.

**Comment by data-man** [comment]
> Already fixed in LLVM master, no idea what commit fixed it though.

Proposal: add field `llvm_version` somewhere.

File: test/stage1/behavior/vector.zig (hunk 1 above)
Comment (data-man) [reply test/stage1/behavior/vector.zig:498]: > If you don't believe me

It is impossible. :)

> you'll see that the test will pass with no problem at all.

Tests passed on my PC.

Comment (LemonBoy) [reply test/stage1/behavior/vector.zig:498]: Any test producing a NaN will pass as the comparisons evaluate to false, if you remove this lines the test is not checking anything.

**Comment by LemonBoy** [comment]
> Remove min/max tests for NaNs 

Why?

**Comment by data-man** [comment]
@LemonBoy 
> Why?

>lib/std/debug.zig:231:69: error: cannot store runtime value in compile time variable
const first_trace_addr = if (builtin.os.tag == .wasi) null else @returnAddress();
^
./vector.zig:500:40: note: called from here
std.debug.panic("unexpected NaN value!\n", .{});
^
./vector.zig:593:28: note: called from here
doTheTestReduce(.Min, [4]f16{ -1.9, 5.1, f16_nan, 100.0 }, f16_nan);

**Comment by LemonBoy** [comment]
The first thing to do when you see a test fail is **to** :clap: **find** :clap: **out** :clap: **what** :clap: **the** :clap: **problem** :clap: **is** :clap: and _not_ to disable them.

That's a bug in the comptime evaluation of min/max and I'm about to send a PR to fix it and, to top everything off, there may be a bug in LLVM that I haven't yet reduced.

**Comment by data-man** [comment]
Oh, I thought the first thing was to cover all possible cases. ;)

**Comment by LemonBoy** [comment]
Ok, it turns out that horizontal min/max reductions behaviour wrt NaNs in LLVM 11 are underspecified and that's being addressed by a recent set of patches. Even for scalar vectors it sometimes fails in the instruction selection pass, this can be seen by trying to compile this set of tests for `aarch64`...

Feel free to put in a `if (false)` block the min/max test cases on NaN and selectively skip the failing tests. Once that's done we need a tracking ticket to review&check the intrinsic behaviour before the next release.

**Comment by data-man** [comment]
After some investigations:
- `Min/Max` doesn't work only for `i64/u64` in `aarch64`
- `f128` doesn't work in `sparc`

What suggestions?

File: test/stage1/behavior/vector.zig [hunk 2]
```diff
@@ -497,72 +497,112 @@ test "vector reduce operation" {
...
 
+            doTheTestReduce(.Min, [4]i16{ -1, 2, 3, 4 }, @as(i16, -1));
+            doTheTestReduce(.Min, [4]u16{ 1, 2, 3, 4 }, @as(u16, 1));
             doTheTestReduce(.Min, [4]i32{ 1234567, -386, 0, 3 }, @as(i32, -386));
-            doTheTestReduce(.Max, [4]i32{ 1234567, -386, 0, 3 }, @as(i32, 1234567));
+            doTheTestReduce(.Min, [4]u32{ 99, 9999, 9, 99999 }, @as(u32, 9));
 
-            doTheTestReduce(.Add, [4]i32{ -9, -99, -999, -9999 }, @as(i32, -11106));
-            doTheTestReduce(.Add, [4]i64{ 9, 99, 999, 9999 }, @as(i64, 11106));
+            if (std.builtin.arch != .aarch64) {
```
Comment (andrewrk) [inline test/stage1/behavior/vector.zig:532]: Please open an issue to track this test failure and link it here where it is being disabled

File: test/stage1/behavior/vector.zig [hunk 3]
```diff
@@ -497,72 +497,112 @@ test "vector reduce operation" {
...
-
             doTheTestReduce(.Add, [4]f32{ -1.9, 5.1, f32_nan, 100.0 }, f32_nan);
-            doTheTestReduce(.Add, [4]f32{ -1.9, 5.1, f32_nan, 100.0 }, f32_nan);
-
-            doTheTestReduce(.Add, [4]f64{ -1.9, 5.1, f64_nan, 100.0 }, f64_nan);
             doTheTestReduce(.Add, [4]f64{ -1.9, 5.1, f64_nan, 100.0 }, f64_nan);
 
-            doTheTestReduce(.Mul, [4]f16{ -1.9, 5.1, f16_nan, 100.0 }, f16_nan);
-            doTheTestReduce(.Mul, [4]f16{ -1.9, 5.1, f16_nan, 100.0 }, f16_nan);
+            if (false) {
```
Comment (andrewrk) [inline test/stage1/behavior/vector.zig:594]: please open or find the relevant issue and link it here

**Review by andrewrk — CHANGES_REQUESTED** [review]

**Comment by data-man** [comment]
Ready to next review or merge.

File: test/stage1/behavior/vector.zig (hunk 3 above)
Comment (LemonBoy) [reply test/stage1/behavior/vector.zig:600]: This is a different problem than the one above.

LLVM's handling of nans across different arches improved a lot after the 12 release.

**Pull request merged**